MovingBlocks/FacadeServer
# Authentication management

# Overview

Here is the code which implements user authentication via HTTP and WebSocket as described in the [devlog updates](https://gianluca-nitti.github.io/GSoC-2017-devlog/) for weeks 2-3.

Here is a little summary of the updates (more details in the posts at the link above, and in the recent messages on Slack#server-api):
- This facade will provide access to various resources of a running Terasology server via HTTP and WebSocket clients, exchanging messages in the JSON data serialization protocol.
- The authentication protocol is designed to resemble the already existing one for regular Terasology client, to achieve a good security level. This means that an authentication handshake roughly happens in this way:
  - The client asks to authenticate
  - The server sends an hello message which contains it's public certificate along with some random data and a timestamp
  - The client **must** reply with another hello message composed in the same way (but using the client public certificate instead of the server one), followed by a verification payload which is obtained by concatenating the server hello (received) with the client hello (generated and sent) and RSA signing the result using the client private certificate as key.
  - The server verifies the signature, and if it's valid the client is authenticated.
  - On successful authentication, the server also sends a verification payload constructed in the same way but signed with the server private certificate
  - *Optionally*, the client can verify the server identity by verifying the signature against the server public certificate, and optionally disconnect if this verification fails and since the server is not who it claims to be
- For WebSocket, the handshake is simply a sequence of messages on the same socket; for HTTP, every "ping-pong" (e.g. "init->obtain server hello", "send client hello and signed data->receive server signed data or authentication failure") is a "request-response" pair; on the first request, the server will also send a session token as a response header and the client must include it as a request header for the other requests.

# How to test
There is a simple Javascript WebSocket client as a proof-of-concept that the RSA part can be handled by a browser (with the right libraries) [here](https://gist.github.com/gianluca-nitti/14e5d61d66d1f347330fa7f42f3db7da). For HTTP I tested manually using [Postman](https://www.getpostman.com/).

Note that the BigIntegers in the certificates are serialized as a base64 string representing their binary representation; **the [BigIntegerBase64Serializer](https://github.com/gianluca-nitti/Terasology/blob/09273389852b8573c3c0d2f1811f2dfb80d0dabd/engine/src/main/java/org/terasology/identity/storageServiceClient/BigIntegerBase64Serializer.java) class implemented in MovingBlocks/Terasology#2965, thus this branch will not compile if the root repository (where this facade is located as a nested repository in facades/Server) is not at that branch or that PR is merged**.
The reasons behind this is that handling very large numbers in JSON is unhandy, for example in Javascript external libraries must be used, and that by using base64 the messages are shorter (a little bandwidth is saved, but probably this isn't very relevant). This also means that directly pasting the "normal" configuartion file in the simple client linked above won't work, it needs to be re-encoded with the BigIntegers in the correct format. The hack I used was to start the regular client in debug mode with a breakpoint on the code that saves the configuration, and then using the IntelliJ debugger's "evaluate expression" feature to add the correct gson type adapter and obtaining a JSON string of the configuration with the numbers in the right format. Since the Java's BigInteger binary representation is simply [the two's complement representation of the number](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#toByteArray--) manually converting them with a tool [like this](https://conv.darkbyte.ru/) (pasting a number in the DEC field and getting the base64) should also work but I have not tried it yet.
Any suggestion is welcome, maybe another version of `BigIntegerBase64Serializer` in this repository should be made so that the server could handle both the representations?
